README
Notes about the Clock activity

License
=======
This activity and its code is put under public domain. Do whatever you want with it!
If you want to contact me: Pierre Métras <pierre@alterna.tv>


Localization of the Clock
=========================
The code for the Clock activity tries to be easily localizable to be used in Sugar with different languages and locales. To achieve that, we tried to put into gettext dictionnary all elements which could be culturally different, like the fonts used to display texts in the clock. We used also Pango to have a better font management and give the translators the ability to adapt the display of text to their locale.

The Clock uses small messages. As with GUI software, it could be that the same messages have to be used in different contexts, and sometimes localized with different values (depending of the context). For instance, the font used to print the date in the Analog and Digital clock views are the same in the English/USA environment, but the font for the Digital one has to be reduced when using the Indi environment.
GNU gettext tool provides the pgettext() function to deal with such situations (see http://www.gnu.org/software/gettext/manual/gettext.html#Ambiguities). Saddly, the Python binding of gettext does not offer this function. I had to define a custom pgettext() and use it as a workaround, while we wait to have it included in a future Python version.


Note to translators: how to translate Pango markup?
---------------------------------------------------
Usage of Pango markup gives you a great deal of flexibility in the localization of the Clock activity. Let's take an example from the po/fr.po French localization file:

# TRANS: datetime.strftime() format
#, no-python-format
msgctxt "Digital Clock"
msgid "<markup><span lang=\"en\" font_desc=\"Sans Bold 80\">%x</span></markup>
msgstr "<markup><span lang=\"fr\" font_desc=\"Sans Bold 70\">%d %B %Y</span></markup>

First, I changed the language code from "en" to "fr". Using the right language code let Pango determine the right rules for the layout of the text, for instance from right to left generally for Arabic.

Second, the "%x" format was not correctly supported by the internalization libraries on Build 653: it printed the date as "2008-03-26" instead of "26/03/2008" in French usage. So I decided to use a custom date layout to work around that bug. While at it, I decided that I would display the month name in full letters instead of the number of the month, like "26 mars 2008", with the strftime format "%d %B %Y".

After having changed the date format, I felt that some dates, like "31 octobre 2008" won't display correctly with all display orientations, and I went to reduce the font size from 80 pt to 70 pt. You could also decide to change the font family and use a better font instead of "Sans".

Don't hesitate to adapt the Clock to your local needs...


How to write the time in full letters for your locale?
------------------------------------------------------
The activity use a small inference engine to translate a numeric time to words. This translation is done following rules which have to be localized. I've provided 3 examples, for English, French and Spanish.
This time in full letters is then used by the Text To Speech engine espeak to read aloud the time on demand or when minutes change.

The inference engine tries to find a rule whose pattern is "time(h,m)" where the variable "h" is bound to the current hours and the variable "m" is bound to then current minutes. The rules are scanned in the order they are written, until we find one matching.

For the match to succeed, we can add range conditions with the syntax "[val1 < val2]" or "[val1 < val2 < val3]" after the rule pattern. For instance, if we want to write two rules for hours in the morning and the afternoon, we could write:
	time(h, m) [h < 12] => ... |
	time(h, m) [11 < h] => ... |

As you can see from the previous example, the conditions of the rules precedes the sign "=>" which specifies the right side of the rules with the actions to execute when the match on the left side succeeds. The rules ends at the "|" character.

On the right side of a rule, after the "=>" sign, you can put a combination of text or other rules calls. The text is used as-is and substituted in the result. If it's a rule call, the inference engine will try again from the start to find a matching rule. For instance:
	time(12, 0) => noon |
	time(24, 0) => midnight |
	time(h, 0) => number(h) o'clock |
With the first two rules, the engine will display "noon" or "midnight" when called respectively with the values "time(12, 0)" or "time(24, 0)". With the last rule, when called with the pattern "time(h, 0)" where "h" is value different from 12 or 24 (because caught by the two previous rules), the engine will try to find a rule name "number(_)" and when found and matched, will replace the value returned in the text to be concatenated with the text "o'clock".

Inside right side of rules, that is between "=>" and "|", spaces are significative. You can use the operator "#" to concatenate two pieces of text or rules calls, like in:
	hour(h) => number(h) hour#plural(h) |
	plural(1) => |
	plural(_) => s |
Use "\#" to get the character "#". This is usefull when using pango markup in the text.

As you have probably understod from now, rules patterns or calls use functional syntax: the name of the rule, with arguments between parenthesis. Numerical arguments are considered constants (could be hours or minutes, for example), while alphabetic are variables, bould during the pattern matching or call. The special variable "_" (underscore) is a anonymous variable: its value is not important to the rule firing. For instance, in the previous example, then rule "plural(_) => |" can be read "the plural of anything which is not 1 is 's'".

To test your set of rules, you can create the file 'test_timewriter/LANG_rules.py' where LANG is the ISO code for your language. It's simpler to copy an existing file from the test_timewriter directory and adapt it to your language.
In that file, you just have to declare the '_time_rules' variable containing the source code of your rules. For instances, something like the following:
	#! /usr/bin/env python
	# -*- coding: utf-8 -*-
	_time_rules = """
	    time(h, m) => ...
        """
Then, you can check your rules with the command:
	$ python timewriter.py LANG
where LANG is the ISO code for your language. It will display all the rules analyzed by the timewriter parser, and then display all times from 0:00 to 23:59.


Pango markup in the rules
-------------------------
As this activity is aimed at learning how to read the time, we have to offer ways for the children to recognize the important parts of the time expression. The use of a consistent color conventions is part of the job. Hours are displayed in blue, minutes in green, seconds in red as frequently seen in analog clock faces.

This color convention has to be kept, whenever possible, in the time displayed in full letters. The colors have been selected to maximize contrast on the XO. It means that the inference rules used to write the time must include pango markup around the different parts of the time in the sentence, like:

	 time(h, m) [m < 31] => <span foreground="\#00B20D">min(m)</span> past <span foreground="\#005FE4">hour(h)</span> am_pm(h) |

As this reduce the readability of the rules, I suggest you to write the set of localized rules without the coloring, and later add the pango codes.

The color codes to use:
- Hours		blue: #005FE4
- Minutes	green: #00B20D
- Seconds	red: #E6000A (You probably never have to use this one)
- Days		dark red: #B20008
- Months	purple: #5E008C
- Years		brown: #9A5200


How to debug your localization file?
------------------------------------
As you've seen from previous paragraphs, a lot of parameters and logic for the activity is taken from the localization file. It could happen that a change or translation you've done breaks the activity.
To find where the problem lies, it helps to start the Log activity and look at the log file tv.alterna.Clock-N.log. You could find some hints where Python or Sugard breaks when running the activity.
You can also send me an email for assistance, to <pierre@alterna.tv>. Don't forget to attach your locale file and some explanations where it breaks.


Commands used to generate localized messages files
==================================================
CAREFULL: _p() Python method is mapped to pgettext(msgctxt, msgstr).

The following procedure has to be done from a XO or a computer using UTF-8 as system encoding.

Clock$ is the prompt when I'm in the Clock directory.

First run
---------
# Create the first messages template
Clock$ xgettext --output=po/Clock.pot --add-comment --keyword=_p:1c,2 clock.py timewriter.py speaker.py

# Adapt po/Clock.pot file with author, copyright, etc.
Clock$ vi po/Clock.pot

# Initialize a new locale for French
Clock$ msginit --input=po/Clock.pot --output-file=po/fr.po --locale=fr

# Translate the messages in French
Clock$ vi po/fr.po

# Create the binary messages file
Clock$ msgfmt --output-file=locale/fr/LC_MESSAGES/tv.alterna.Clock.mo --check po/fr.po

Messages update
---------------
When the sources have changed

# Extract the new messages from the source like upper, in the po/Clock.pot file.

# Update a translated message file
Clock$ $ msgmerge --update --backup=simple po/fr.po po/Clock.pot

# Translate the new messages and generate the binary file.



BUGS
====
- SOLVED: Python does not support pgettext (particular gettext). So it makes it difficult to localize identical short strings used in different contexts. clock.py includes a custom function _p(). Better localization support has been added to Python 2.6, in particular pgettext.

- SOLVED: Python 2.5 as used on OLPC XO has a bug with datetime.strftime(pattern), when pattern is longer than around 100 characters. time.strftime() does not exhibit that bug and is used as a workaround. Python 2.5.1 seems not to have this bug. http://bugs.python.org/issue2490.
Version 5 of the Clock activity was developped specifically for XO build 8.2.0 which solved this bug. Use version 4 of the Clock with older builds.

- Whenever the label where the time is displayed becomes wider than the screen width, forinstance when you rotate the display, it would be nice to have line wrapping. I was unable to get it working. The Label.set_line_wrap(True) method does not give the expected result.

- SOLVED: The first time the talking clock or the "time in letters" toolbar button ispressed, their is a few seconds latence when Python loads the code. Some can think that these buttons have no action; please wait a few seconds... This effect has been lowered by using threads in GObject library.


POINT OF INTEREST IN THE CODE
=============================
- Developped at 99% on a XO. I was only missing a SVG editor for the icons. But even the icons where hand crafted...

- How to create a new GTK+ widget and paint its content.

- The use of pango markup to offer better flexibility in localization.

- The signal "notify::active" sent by Sugar to an activity when it comes on the front. In this activity, we stop refreshing the display when we are not active.

- Full pydoc documentation.

- Using the badly documented gobject.threads_init() to have better messages processing.


TODO
====
A drop box for ideas. While the goal is to keep the source simple and understandable by Python starters, here are listed some ideas of features to include in the Clock activity to help learning how to read the time and have at the same time a usefull clock.
- An alarm clock mode
- A stopwatch mode
- Print the timezone
- A talking clock: speaking the time aloud every minute. DONE in version 5.
- A Big Ben chime...
- Add a second toolbar to print calendars and to learn how to read the date.


NEWS
A simple Clock face to know the time of the day, and eventually learn how to read it.

20080129 1.0	- Initial creation of the Basic Clock.
		- 3 display modes: simple, with hours ticks and date, nice clock face.

20080321 2.0	- Added a digital clock with two views: basic and complete.
		- Support context in gettext with custom function pgettext().
		- Changed fonts as they could not display some characters in Portuguese. Use Pango to layout text now.
		- Added a color code to help learning time. Made clock hands a bit wider too.
		- A few bugs solved and made code more complex to read!

20080923 3.0	- Added license information required by the new 8.2 activity bundle format.

20080925 4.0	- When the activity is not active, we don't update the clock to save CPU resources.

20081022 5.0	- Refactoring of the code: cairo drawing has been removed and we use Gdk now. The big drawing method has been splitted in small and easy to understand methods.
		- Display the date and weekday combined out of the clock face.
		- Use XO approved colors for better contrast.
		- New toolbar instead of the hidden right button. Only 3 display modes have been kept for the clock faces: simple, nice and digital.
		- New activity icon.
		- Write the current time in full letters, with the same colors code for the clock hands.
		- Inference engine to translate a numeric time to letters. Three set of localized rules, for English, French and Spanish have been provided as model for translatorsand procedures to test them.
		- Speak aloud the current time through espeak TTS.
		- Most methods are now private.


timewriter.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-
#
# Code released in the Public Domain. You can do whatever you want with this package.
# Look at README file to see how to adapt this program.
# Originally written by Pierre Métras <pierre@alterna.tv> for the OLPC XO laptop.


"""Inference engine to write times in English (or other language).

Change the rules to adapt to another language. Caution: the rule parser does not
strictly check the syntax and many errors will remained ignored.

Usefull functions:
- print_rules(): Dump the set of rules.
- test_times(): Try to print all times from 00:00 to 23:59.
- write_time(hour, minute): Write the (hour, minute) in natural language.
- eval_rule(text): Translate the text according to the set of rules.

Example of usage:
-----------------
import timewriter

w = timewriter.TimeWriter("en")
s = w.write_time(2, 33)
print "It is %s."

prints --> It is thirty-three minutes past two in the morning.

How to tell time in English?
----------------------------
http://en.wikipedia.org/wiki/12-hour_clock
"""

#import pdb

import tokenize
import cStringIO
import re
import copy

from gettext import gettext as _

"""
Grammar for the rules:
----------------------
Root := Pattern
Pattern := Text? Pattern_call? Pattern?
Pattern_call := Pattern_name ( Argument [, Argument]* )
Text := string
Pattern_name := string
Argument := Dumb_variable_ | Variable | Value
Dumb_variable := '_'
Variable := string
Value := number
Rules := Rule ('|' Rules)*
Rule := Pattern_call Range_condition* '=>' Text? Pattern_call? Pattern?
Range_condition := '[' Argument '<' Argument ('<' Argument)? ']'

'#' can be used to concatenate two Texts or Pattern_call without a space between.
"""


class _Rule:
    """A rule is composed of conditions and a body.
        Rule: Conditions => Body

    Examples:
        number(8) => eight
        plural(1) =>
        plural(_) => s

    The inference engine tries to match a template with the conditions pattern of
    the rule, eventually binding the variables. If the rule matches, then the body
    of the rule is used, after substituting the variables by their values and
    eventually firing the other rules called in the body definition.
    """

    def __init__(self, pattern, ranges, body):
        """Create a new rule from its conditions, optional ranges and body.
        """
        self._pattern = pattern
        self._ranges = ranges
        self._body = body


    def get_pattern(self):
        """Gets the conditions pattern of the rule.
        Returns a list [rule_name, arg1, arg2...].
        """
        return self._pattern


    def get_ranges(self):
        """Gets the range condition to apply the rule.
        Returns a list [[arg1, arg2], [arg1, arg2, arg3]...].
        """
        return self._ranges


    def get_body(self):
        """Gets the body of the rule.
        Returns a list, for instance with two text fragments around another rule
        call, [text1, [rule_name, arg1, arg2], text2].
        """
        return self._body


    def __str__(self):
        """Gets the external representation of the rule as lists.
        """
        return "Rule: %s %s => %s" % (self._pattern, self._ranges, self._body)


    def __repr__(self):
        """
        Returns the external representation of the rule.
        """
        return self._repr_call(self._pattern) + self._repr_ranges(self._ranges) + " => " + self._repr_body(self._body)


    def _repr_call(self, call):
        """Returns the external representation of a rule call.
        """
        return "%s(%s)" % (call[0], ", ".join(str(x) for x in call[1:]))


    def _repr_ranges(self, ranges):
        """Returns the external repressentation of a rule ranges.
        """
        result = ""
        for r in ranges:
            if len(r) == 2:
                result += " [ %s < %s ]" % (r[0], r[1])
            else:
                result += " [ %s < %s < %s ]" % (r[0], r[1], r(2))
        return result


    def _repr_body(self, body):
        """Returns the external representation of a rule body.
        """
        result = ""
        for item in body:
            if isinstance(item, list):
                result += "#" + self._repr_call(item)
            else:
                result += item
        return result


class TimeWriter:
    """A class to print the time in natural language.
    """


    def __init__(self):
        """Create a time writer for the current language.
        The rules localized for a language are stored in the localized messages file.
        """
        self._rules = self.parse_rules(self._time_rules)


    # TRANS: The rules to print the time in the localized language.
    #
    # Example syntax:
    #     time(h, 15) => a quarter to hour(h) am_pm(h) |
    # The left hand side of the rule defines a pattern with a variable 'h' and a
    # value '15'.
    # The right hand side, when applied, will use the text "a quarter to " and call
    # the first rule matching hour(h) after substituting the variable 'h' by its value,
    # and call the rule matching am_pm(h).
    # Internal spaces are significant on the right side of a rule. In calls, all
    # arguments which are not numbers are considered to be variables. The rule parser
    # is very simple and will let many syntax errors go ignored.
    #
    # A rule ends with the character '|'.
    # The character '_' is a anonymous variable.
    # The character '#' can be used to concatenate two text fragments. For instance:
    #     plural(1) => |
    #     plural(_) => s |
    #     hour(h) => number(h) hour#plural(h) |
    # Use '\#' to use a # character, for instance in a pango color
    # tag like <span foreground="\#FF0055">
    #
    # You can put range conditions on firing a rule, with the syntax [var1 < var2] or
    # [var1 < var2 < var3]. For instance:
    #     hours(h) [h < 12] => in the morning |
    #     hours(h) [12 < h < 18] => in the afternoon |
    #     hours(_) => in the night |
    #
    # These rules will be called with the root pattern "time(hour, minute)", with the
    # variable 'hour' bound to the current hour and the variable 'minute' to the
    # current minute.
    # Order of rules is important. Rules are tried from first to last. So most precise
    # rule must be placed first in the list.
    #
    # You can validate your set of rules by running the command line:
    #     python timewriter.py LANG
    #
    # You should use pango markup to respect the same colors as for the clock hands.
    # Look at the README file from the activity for explanations on how to create
    # rules.
    _time_rules = _("""time(h, m) => What Time Is It?""")


    def _syntax_error(self, rule):
        """Print an error message when a rule can't be parsed.
        """
        raise SyntaxError("Syntax error in rule: %s" % rule)
    
    
    def print_rules(self):
        """Print the list of rules. Can be used to check the parser.
        """
        print "Rules = ["
        for i, rule in enumerate(self._rules):
            print "#%d %s" % (i, rule)
        print "]\nTotal = %d rules\n" % len(self._rules)
    
    
    def repr_rules(self):
        """Gets the external representation of the rules.
        """
        return " |\n".join(repr(rule) for rule in self._rules)
    
    
    def parse_rules(self, source):
        """Parse all the rules for the current language.
        Rules are a list of rule definitions separated by |.
            Rules := Rule ( '|' Rule )*
        Returns the list of rules.
        """
        self._rules = []
        for rule in source.split("|"):
            r = self._parse_rule(rule)
            self._rules.append(r)
        return self._rules
    
    
    def _parse_rule(self, source):
        """Parse a single rule.
        A rule is composed of a pattern and a body, separated by =>.
            Rule := Pattern_call Range_condition* '=>' Rule_body
        Return a rule definition object.
        """
        r = re.findall(r"\s*(\w+\s*\(.*?\))\s*(\[.*\])?\s*=>(.*)", source)
        if r[0] is None or r[0][0] == "" or r[0][2] == "":
            self._syntax_error(rule)
        pattern_call = self._parse_call(r[0][0])
        range_conditions = self._parse_ranges(r[0][1])
        rule_body = self._parse_body(r[0][2].strip())
        rule = _Rule(pattern_call, range_conditions, rule_body)
        return rule
    
    
    def _parse_call(self, source):
        """Parse a rule pattern or call.
        A rule call is similar to a function call.
            Rule_call := Rule_name '(' ( arg [',' arg]* ) ')'
        Returns a list [Rule_name, arg1, arg2...]
        """
        src = cStringIO.StringIO(source).readline
        src = tokenize.generate_tokens(src)
        token = src.next()
        if token[0] is not tokenize.NAME:
             self._syntax_error(source)
        call = [token[1]]
        token = src.next()
        if token[1] != "(":
             self._syntax_error(source)
        token = src.next()
        while token[1] != ")":
            try:
                call.append(int(token[1]))
            except ValueError:
                call.append(token[1])
            token = src.next()
            if token[1] == ",":
                token = src.next()
        return call


    def _parse_ranges(self, source):
        """Parse zero or many range conditions.
            Range_conditions := Range_condition*
            Range_condition := '[' arg1 '<' arg2 ('<' arg3)? ']'
        Returns a list [[arg11, arg12], [arg21, arg22, arg23]...]
        """
        if source == "":
            return None
        else:
            ranges = []
            for r in re.findall(r"\[\s*(.*?)\s*<\s*(.*?)\s*(?:<\s(.*?))?\s*\]", source):
                rang = []
                for x in r:
                    if x != "":
                        try:
                            rang.append(int(x))
                        except ValueError:
                            rang.append(x)
                ranges.append(rang)
            return ranges


    def _parse_body(self, source):
        """Parse the right hand side of a rule.
        We must preserve spaces in the rule body, so we use regular expression for parsing.
            Rule_body := text? Pattern_call? Rule_body?
        Returns a list [text, (pattern, arg1, arg2...), text, ...]
        '#' is a concatenation operator if not escaped by '\'
        """
        if not re.search(r"(\w+\s*\(.*?\))", source):
            # No rull call present
            return [source]
        else:
            body = []
            text = ""
            escaped = False
            for item in re.findall(r"(?:(\w+\s*\(.*?\))|(.))", source):
                if item[0] != "":
                    if text != "":
                        body.append(text)
                        text = ""
                    body.append(self._parse_call(item[0]))
                else:
                    if item[1] == "\\":
                        escaped= True
                    elif item[1] == "#":
                        if escaped:
                            text += item[1]
                            escaped = False
                    else:
                        text += item[1]
                        escaped = False
            if text != "":
                body.append(text)
            return body
    
    
    def write_time(self, hour, minute):
        """Gives the natural language translation of the time.
        For instance, write_time(3, 41) returns "three hours and forty-one minutes in the morning" with an English TimeWriter.
        """
        return self.eval_rule("time(%s, %s)" % (hour, minute))


    def eval_rule(self, source):
        """Evaluate the source against the set of rules.
        Example: eval_rule("It is time(15, 2).")
        """
        lst = self._parse_body(source)
        # lst = [text, [call, arg1, arg2..], text, ...]
        # The goal is now to flatten the list lst resolving all the calls
        lst = self._match_pattern(lst)
        return "".join(lst)
    
    
    def _match_pattern(self, patterns):
        """Match a list of patterns agains the set of rules.
        This engine stops at the first rule matching the pattern and eventually
        instanciates the variables, then recursively apply them in the body of the
        matched rule.
        Returns a list with all the patterns replaced by the rules bodies infered.
        If a pattern can't be matched, the engine produces no result in the resulting
        list. As we expect the set of rules to be complete (all submitted patterns
        fire at least one rule), we raise an exception if the number of items in the
        result is not the same as the number of patters submitted.
        """
        result = []
    
        for pattern in patterns:
            if isinstance(pattern, list):
                for rule in self._rules:
                    cond = rule.get_pattern()
    
                    # Simple tests first
                    if len(pattern) != len(cond):
                        continue
                    # Check that we test the same set of rules
                    if pattern[0] != cond[0]:
                        continue
    
                    # We use lazy rule body copy, only when there is a variable in the
                    # pattern. I've found that I got a 50% boost in performance doing
                    # that instead of doing the deepcopy from start.
                    body = None
    
                    # Now we check that the premises match and substitute all
                    # variables in rule body.
                    match = True
                    # The dictionary will keep the variable bindings
                    bind = {}
                    for i in range(1, len(pattern)):
                        # Dumb variable
                        if cond[i] == "_":
                            continue
    
                        # Variable instanciation
                        if not isinstance(cond[i], int):
                            if body is None:
                                body = copy.deepcopy(rule.get_body())
                            body = self._apply_var(cond[i], pattern[i], body)
                            bind[cond[i]] = pattern[i]
                            continue
    
                        if pattern[i] != cond[i]:
                            match = False
                            break

                    # Checking the range conditions to see if we can apply the rule
                    ranges = rule.get_ranges()
                    if match and ranges:
                        ranges = copy.deepcopy(ranges)
                        for r in ranges:
                            # Bind all variables
                            for i in range(0, len(r)):
                                if not isinstance(r[i], int):
                                    try:
                                        r[i] = bind[r[i]]
                                    except KeyError:
                                        #self._syntax_error(rule)
                                        match = False
                                        break

                            # Now check that the range is valid
                            if r[0] >= r[1]:
                                match = False
                                break
                            if len(r) > 2 and r[1] >= r[2]:
                                match = False
                                break

                    if match:
                        # Then we apply all the rule calls in the body of the rule
                        # that matched
                        if body is None:
                            body = copy.deepcopy(rule.get_body())
                        calls = True
                        while calls:
                            calls = False
                            for i in range(0, len(body)):
                                if isinstance(body[i], list):
                                    body[i] = "".join(self._match_pattern([body[i]]))
                                    calls = True
                        result.append("".join(body))
                        break
            else:
                result.append(pattern)
    
        if len(result) != len(patterns):
            raise Exception("There is a missing rule; match failed for pattern %s..." % patterns)
    
        return result


    def _apply_var(self, var, value, body):
        """Instanciate a variable in the body of a rule.
        Returns the body of the rule with that variable substituted by its value in all calls.    This function eventually changes the 'body' argument.
        """
        for elem in body:
            if isinstance(elem, list):
                for i in range(1, len(elem)):
                    if elem[i] == var:
                        elem[i] = value
        return body


    def test_times(self):
        """Check that the time rules are complete, printing all combinations.
        """
        print "***** Checking all times *****"

        for h in range(0, 24):
            for m in range(0, 60):
                str = self.write_time(h, m)
                print "time(%d, %d) -> %s" % (h, m, str)


    def set_rules(self, rules_source):
        """Assign the source of rules to the timewriter instance.
        The rules are parsed during the operation.
        """
        self._rules = self.parse_rules(rules_source)



def main():
    """Main entry point to test rules.
    """
    # import sys
    if len(sys.argv) == 1:
        print "Usage: python timewriter.py lang"
        print "Where lang is a ISO language code (en, fr, es...)"
        print "TimeWriter rules must be available in directory test-timewriter."
        exit(1)
    lang = sys.argv[1]
    test_mod = "test_timewriter." + lang + "_rules"
    import_mod = "from " + test_mod + " import _time_rules as test_rules"
    exec import_mod
    w = TimeWriter()
    w.set_rules(test_rules)
    print "***** Rules parsed *****"
    w.print_rules()
    w.test_times()


# Run "$ python timewriter.py en" to check all rules for English ("en" argument)
# Run "$ python timewriter.py en 1" to get profiling information.
if __name__ == "__main__":
    import sys
    if len(sys.argv) > 2:
        import cProfile
        cProfile.run("main()", "genprof")
        import pstats
        p = pstats.Stats("genprof")
        print
        print "***** Profiling *****"
        p.strip_dirs().sort_stats("time", "name").print_stats(0.1)
    else:
        main()

